File: app/src/main/java/bb/incognito/view/adapter/GuestAdapter.java
package bb.incognito.view.adapter;

import android.databinding.DataBindingUtil;
import android.databinding.adapters.AdapterViewBindingAdapter;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Filter;
import android.widget.Filterable;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import bb.incognito.R;
import bb.incognito.model.Guest;
import bb.incognito.model.GuestWithCocktails;
import bb.incognito.repositories.GuestWithCocktailsRepository;
import bb.incognito.viewModel.GuestRowVM;
import bb.incognito.databinding.GuestRowBinding;

public class GuestAdapter extends RecyclerView.Adapter<GuestAdapter.GuestViewHolder> implements Filterable {
    private GuestWithCocktailsRepository guestWithCocktailsRepository;
    private List<GuestWithCocktails> guests;
    private List<GuestWithCocktails> filteredGuests;
    private ItemFilter filter = new ItemFilter();

    private ItemTouchHelper.Callback callback = new ItemTouchHelper.Callback() {
        @Override
        public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
            final int dragFlags = 0;//ItemTouchHelper.UP | ItemTouchHelper.DOWN;
            final int swipeFlags = ItemTouchHelper.START | ItemTouchHelper.END;
            return makeMovementFlags(dragFlags, swipeFlags);
        }

        @Override
        public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
            return true;
        }

        @Override
        public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
            GuestWithCocktails g =  filteredGuests.get(viewHolder.getAdapterPosition());

        }
    };

    public ItemTouchHelper itemTouchHelper = new ItemTouchHelper(callback);

    public void setGuests(List<GuestWithCocktails> guests) {
        this.guests = guests;
        this.filteredGuests = guests;
        notifyDataSetChanged();
    }

    @Override
    public GuestViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        GuestRowBinding cardRowBinding =
                DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.guest_row,
                        parent, false);
        return new GuestViewHolder(cardRowBinding);
    }

    @Override
    public void onBindViewHolder(GuestViewHolder holder, int position) {
        if (filteredGuests != null)
        {
            holder.bindCard(filteredGuests.get(position));
            holder.itemView.setOnContextClickListener(view -> {
                itemTouchHelper.startSwipe(holder);
                return false;
            });
        }
        else
        {
            holder.bindCard(new GuestWithCocktails(new Guest("Nie ma gości :(")));
        }
    }

    @Override
    public int getItemCount() {
        return filteredGuests != null ? filteredGuests.size() : 0;
    }

    @Override
    public Filter getFilter() {
        return filter;
    }

    class GuestViewHolder extends RecyclerView.ViewHolder{
        GuestRowBinding guestRowBinding;


        public GuestViewHolder(GuestRowBinding guestRowBinding) {
            super(guestRowBinding.guestRow);
            this.guestRowBinding = guestRowBinding;
        }

        void bindCard(GuestWithCocktails guest) {
            if (guestRowBinding.getViewModel() == null) {
                guestRowBinding.setViewModel(
                        new GuestRowVM(guest));
            } else {
                guestRowBinding.getViewModel().setGuest(guest);
            }
        }
    }

    private class ItemFilter extends Filter {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            FilterResults results = new FilterResults();
            List<GuestWithCocktails> newList = new ArrayList<>(guests.size());

            for(GuestWithCocktails g : guests)
            {
                if(g.getName().toLowerCase().contains(constraint))
                    newList.add(g);
            }
            results.values = newList;
            results.count = newList.size();
            return results;
        }

        @SuppressWarnings("unchecked")
        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            filteredGuests = (ArrayList<GuestWithCocktails>) results.values;
            notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/bb/incognito/view/adapter/GuestAdapter.kt
package bb.incognito.view.adapter

import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable

import java.util.ArrayList

import bb.incognito.R
import bb.incognito.viewModel.GuestRowVM
import bb.incognito.databinding.GuestRowBinding
import bb.incognito.model.Guest
import bb.incognito.model.GuestWithCocktails

class GuestAdapter : RecyclerView.Adapter<GuestAdapter.GuestViewHolder>(), Filterable {
    private var guests: MutableList<GuestWithCocktails>? = null
    private var filteredGuests: MutableList<GuestWithCocktails>? = null
    private val filter = ItemFilter()

    private val callback = object : ItemTouchHelper.Callback() {
        override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
            val dragFlags = 0//ItemTouchHelper.UP | ItemTouchHelper.DOWN;
            val swipeFlags = ItemTouchHelper.START or ItemTouchHelper.END
            return ItemTouchHelper.Callback.makeMovementFlags(dragFlags, swipeFlags)
        }

        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            return true
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
            //val g = filteredGuests!![viewHolder.adapterPosition]
        }
    }

    var itemTouchHelper = ItemTouchHelper(callback)

    fun getGuests(): MutableList<GuestWithCocktails> {
        return guests!!
    }
    fun setGuests(guests: MutableList<GuestWithCocktails>?) {
        this.guests = guests
        this.filteredGuests = guests
        notifyDataSetChanged()
    }

    fun restoreItem(removedPosition: Int, guestWithCocktails: GuestWithCocktails) {
        guests!!.add(removedPosition, guestWithCocktails)
        notifyItemInserted(removedPosition)
    }

    fun removeAt(position: Int) {
        guests!!.removeAt(position)
        notifyItemRemoved(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GuestViewHolder {
        val cardRowBinding = DataBindingUtil.inflate<GuestRowBinding>(LayoutInflater.from(parent.context), R.layout.guest_row,
                parent, false)
        return GuestViewHolder(cardRowBinding)
    }

    override fun onBindViewHolder(holder: GuestViewHolder, position: Int) {
        if (filteredGuests != null) {
            holder.bindCard(filteredGuests!![position])
            holder.itemView.setOnContextClickListener { _ ->
                itemTouchHelper.startSwipe(holder)
                false
            }
        } else {
            holder.bindCard(GuestWithCocktails(Guest("Nie ma gości :(")))
        }
    }

    override fun getItemCount(): Int {
        return if (filteredGuests != null) filteredGuests!!.size else 0
    }

    override fun getFilter(): Filter {
        return filter
    }

    inner class GuestViewHolder(var guestRowBinding: GuestRowBinding) : RecyclerView.ViewHolder(guestRowBinding.guestRow) {

        fun bindCard(guest: GuestWithCocktails) {
            if (guestRowBinding.viewModel == null) {
                guestRowBinding.viewModel = GuestRowVM(guest)
            } else {
                guestRowBinding.viewModel?.setGuest(guest)
            }
        }
    }

    private inner class ItemFilter : Filter() {
        override fun performFiltering(constraint: CharSequence): Filter.FilterResults {
            val results = Filter.FilterResults()
            val newList = ArrayList<GuestWithCocktails>(guests!!.size)

            for (g in guests!!) {
                if (g.name.toLowerCase().contains(constraint))
                    newList.add(g)
            }
            results.values = newList
            results.count = newList.size
            return results
        }

        override fun publishResults(constraint: CharSequence, results: Filter.FilterResults) {
            filteredGuests = results.values as ArrayList<GuestWithCocktails>
            notifyDataSetChanged()
        }
    }
}
